linkerd/proxy/http/src/h1.rs

```diff
@@ -1,45 +1,142 @@
-    trace!(%uri, "Setting origin form");
-    let mut parts = mem::replace(uri, Uri::default()).into_parts();
-    parts.scheme = None;
-    parts.authority = None;
-    *uri = Uri::from_parts(parts).expect("path only is valid origin-form uri")
+pub type ResponseFuture =
+    Pin<Box<dyn Future<Output = Result<http::Response<Body>, hyper::Error>> + Send + 'static>>;
+
+impl<C, T, B> Client<C, T, B>
+where
+    T: Clone + Send + Sync + 'static,
+    C: tower::make::MakeConnection<T> + Clone + Send + Sync + 'static,
+    C::Connection: Unpin + Send + 'static,
+    C::Future: Unpin + Send + 'static,
+    C::Error: Into<Error>,
+    B: hyper::body::HttpBody + Send + 'static,
+    B::Data: Send,
+    B::Error: Into<Error> + Send + Sync,
+{
+    pub(crate) fn request(&mut self, mut req: http::Request<B>) -> ResponseFuture {
+        // Marked by `upgrade`.
+        let upgrade = req.extensions_mut().remove::<Http11Upgrade>();
+        let is_http_connect = req.method() == &http::Method::CONNECT;
+
+        // Configured by `normalize_uri`.
+        let absolute_form = req.extensions_mut().remove::<WasAbsoluteForm>().is_some();
+        debug_assert!(req.uri().authority().is_some());
+
+        let is_missing_host = req
+            .headers()
+            .get(http::header::HOST)
+            .map(|v| v.is_empty())
+            .unwrap_or(true);
+        let rsp_fut = if req.version() == http::Version::HTTP_10 || is_missing_host {
+            // If there's no authority, we assume we're on some weird HTTP/1.0
+            // ish, so we just build a one-off client for the connection.
+            // There's no real reason to hold the client for re-use.
+            debug!(absolute_form, is_missing_host, "Using one-off client");
+            hyper::Client::builder()
+                .pool_max_idle_per_host(0)
+                .set_host(absolute_form)
+                .build(HyperConnect::new(
+                    self.connect.clone(),
+                    self.target.clone(),
+                    absolute_form,
+                ))
+                .request(req)
+        } else {
+            // Otherwise, use a cached client to take advantage of the
+            // connection pool. The client needs to be configured for absolute
+            // (HTTP proxy-style) URIs, so we cache separate absolute/relative
+            // clients lazily.
+            let client = if absolute_form {
+                debug!("Using absolute client");
+                &mut self.absolute
+            } else {
+                debug!("Using relative client");
+                &mut self.relative
+            };
+
+            if client.is_none() {
+                debug!("Caching new client");
+                *client = Some(hyper::Client::builder().set_host(absolute_form).build(
+                    HyperConnect::new(self.connect.clone(), self.target.clone(), absolute_form),
+                ));
+            }
+
+            client.as_ref().unwrap().request(req)
```

Comment:
nit/TIOLI: there's actually an [`Option::get_or_insert_with`] method that basically implements this logic and avoids the unwrap:
```suggestion
            client
                .get_or_insert_with(|| {
                    debug!("Caching new client");
                    hyper::Client::builder().set_host(absolute_form).build(
                        HyperConnect::new(self.connect.clone(), self.target.clone(), absolute_form),
                    )
                })
                .request(req)
```

Reply:
Unfortunately, this doesn't play with the borrow checker and `self` -- because we're holding a mutable ref to the client, we can't pass a reference to self into the closure.